encoding/custom/sema_codec/codec_test.go

```diff
@@ -0,0 +1,1705 @@
+		{sema.InvalidType, sema_codec.EncodedSemaSimpleTypeInvalidType},
+		{sema.MetaType, sema_codec.EncodedSemaSimpleTypeMetaType},
+		{sema.NeverType, sema_codec.EncodedSemaSimpleTypeNeverType},
+		{sema.PathType, sema_codec.EncodedSemaSimpleTypePathType},
+		{sema.StoragePathType, sema_codec.EncodedSemaSimpleTypeStoragePathType},
+		{sema.CapabilityPathType, sema_codec.EncodedSemaSimpleTypeCapabilityPathType},
+		{sema.PublicPathType, sema_codec.EncodedSemaSimpleTypePublicPathType},
+		{sema.PrivatePathType, sema_codec.EncodedSemaSimpleTypePrivatePathType},
+		{sema.StorableType, sema_codec.EncodedSemaSimpleTypeStorableType},
+		{sema.StringType, sema_codec.EncodedSemaSimpleTypeStringType},
+		{sema.VoidType, sema_codec.EncodedSemaSimpleTypeVoidType},
+	}
+
+	for _, typ := range tests {
+		t.Run(typ.SimpleType.Name, func(t *testing.T) {
+			testRootEncodeDecode(t, typ.SimpleType,
+				byte(typ.Type),
+			)
+		})
+	}
+}
+
+func TestSemaCodecNumericTypes(t *testing.T) {
+	t.Parallel()
+
+	type TestInfo struct {
+		SimpleType sema.Type
+		Type       sema_codec.EncodedSema
+	}
+
+	tests := []TestInfo{
+		{sema.NumberType, sema_codec.EncodedSemaNumericTypeNumberType},
+		{sema.SignedNumberType, sema_codec.EncodedSemaNumericTypeSignedNumberType},
+		{sema.IntegerType, sema_codec.EncodedSemaNumericTypeIntegerType},
+		{sema.SignedIntegerType, sema_codec.EncodedSemaNumericTypeSignedIntegerType},
+		{sema.IntType, sema_codec.EncodedSemaNumericTypeIntType},
+		{sema.Int8Type, sema_codec.EncodedSemaNumericTypeInt8Type},
+		{sema.Int16Type, sema_codec.EncodedSemaNumericTypeInt16Type},
+		{sema.Int32Type, sema_codec.EncodedSemaNumericTypeInt32Type},
+		{sema.Int64Type, sema_codec.EncodedSemaNumericTypeInt64Type},
+		{sema.Int128Type, sema_codec.EncodedSemaNumericTypeInt128Type},
+		{sema.Int256Type, sema_codec.EncodedSemaNumericTypeInt256Type},
+		{sema.UIntType, sema_codec.EncodedSemaNumericTypeUIntType},
+		{sema.UInt8Type, sema_codec.EncodedSemaNumericTypeUInt8Type},
+		{sema.UInt16Type, sema_codec.EncodedSemaNumericTypeUInt16Type},
+		{sema.UInt32Type, sema_codec.EncodedSemaNumericTypeUInt32Type},
+		{sema.UInt64Type, sema_codec.EncodedSemaNumericTypeUInt64Type},
+		{sema.UInt128Type, sema_codec.EncodedSemaNumericTypeUInt128Type},
+		{sema.UInt256Type, sema_codec.EncodedSemaNumericTypeUInt256Type},
+		{sema.Word8Type, sema_codec.EncodedSemaNumericTypeWord8Type},
+		{sema.Word16Type, sema_codec.EncodedSemaNumericTypeWord16Type},
+		{sema.Word32Type, sema_codec.EncodedSemaNumericTypeWord32Type},
+		{sema.Word64Type, sema_codec.EncodedSemaNumericTypeWord64Type},
+		{sema.FixedPointType, sema_codec.EncodedSemaNumericTypeFixedPointType},
+		{sema.SignedFixedPointType, sema_codec.EncodedSemaNumericTypeSignedFixedPointType},
+		{sema.Fix64Type, sema_codec.EncodedSemaFix64Type},
+		{sema.UFix64Type, sema_codec.EncodedSemaUFix64Type},
+	}
+
+	for _, typ := range tests {
+		t.Run(typ.SimpleType.String(), func(t *testing.T) {
+			t.Parallel()
```

Comment:
using `t.Parallel` in a sub-test nested in a for-loop doesn't work, see https://github.com/onflow/cadence/pull/1851#discussion_r946218284